File: example/robot.js
import { EventEmitter } from 'events';
import { Motor } from './motor.js';

// For reading and writing to config
import path from 'path';
import fs from 'fs';

// For debugging
import { Debug } from './debug.js';
const logger = Debug('mock:robot' + '\t');

export class Robot extends EventEmitter {
  /** ------------------------------
   * Constructor
   */
  constructor({ id }) {
    logger(`creating robot with id ${id}`);

    // Becasuse we are event emitter
    super();

    this.id = id; // id of the robot
    this.stopped = true; // if the robot is currently stopped
    this.ready = false; // if the robot is ready
    this.homing = false; // if the robot is currently homing
    this.moving = false; // if the robot is moving
    this.motorMap = {}; // tracks all motors by joint id
    this.motors = []; // array of all motors ( used for quick itteration )
    this.errors = []; // array of any errors that got triggered

    // Start up the robot when we are ready ( normally you would be connecting to some control interface and waiting here )
    setTimeout(() => {
      this.setup();
    }, 2000);
  }

  /** ------------------------------
   * setup
   */
  setup() {
    // First read in the config
    this.readConfig();

    logger(`starting robot with id ${this.id}`);

    // Create the motors
    this.motorMap.j0 = new Motor({ id: 'j0', ...this.config.j0 });
    this.motorMap.j1 = new Motor({ id: 'j1', ...this.config.j1 });
    this.motorMap.j2 = new Motor({ id: 'j2', ...this.config.j2 });
    this.motorMap.j3 = new Motor({ id: 'j3', ...this.config.j3 });
    this.motorMap.j4 = new Motor({ id: 'j4', ...this.config.j4 });
    this.motorMap.j5 = new Motor({ id: 'j5', ...this.config.j5 });

    // Set motors into array
    this.motors = Object.values(this.motorMap);

    // Subscribe to events for all motors
    this.motors.forEach((motor) => {
      motor.on('homing', () => this.robotState());
      motor.on('home', (id) => this.motorHomed(id));
      motor.on('disabled', () => this.robotState());
      motor.on('enabled', () => this.robotState());
      motor.on('moved', (id) => this.motorMoved(id));
      motor.on('reset', () => this.robotState());
    });

    // Create Gripper if you have a gripper

    // Report all encoder updates at 100ms interval
    setInterval(() => {
      this.robotEncoder(); // Will emit encoder event
    }, 100);

    this.ready = true;
    this.emit('ready');
  }

  /** ------------------------------
   * readConfig
   */
  readConfig() {
    // Read in config file ( create if it does not exist yet )
    try {
      // Get filename
      const filename = path.resolve('config.json');

      // Check if it exists and create if it does not
      if (!fs.existsSync(filename)) {
        console.log('Config file does not exist creating');
        fs.writeFileSync(
          filename,
          JSON.stringify({
            j0: { limNeg: -180, limPos: 180, homePos: 0 },
            j1: { limNeg: -140, limPos: 80, homePos: 0 },
            j2: { limNeg: -140, limPos: 80, homePos: 0 },
            j3: { limNeg: -180, limPos: 180, homePos: 0 },
            j4: { limNeg: -95, limPos: 95, homePos: 0 },
            j5: { limNeg: -180, limPos: 180, homePos: 0 },
          }),
        );
      }

      // Read in config file
      const config = JSON.parse(fs.readFileSync(filename, 'utf8'));

      logger('Successfully read in config', config);

      this.config = config;
    } catch (err) {
      console.error(err);
    }
  }

  /** ------------------------------
   * writeConfig
   */
  writeConfig() {
    logger('Writing config to file', this.config);
    try {
      // Get filename
      const filename = path.resolve('config.json');
      // Write config
      fs.writeFileSync(filename, JSON.stringify(this.config));
    } catch (err) {
      console.error(err);
    }
  }

  /** ------------------------------
   * updateConfig
   *
   * By default this will NOT save to the file it will only update in memory
   * Note: a call to writeConfig() at any time will save everything that has been updated to the file
   */
  updateConfig(key, value, save = false) {
    logger(`updating config ${key} to ${value}`);

    // Special check ( dont let user set a config param to null !! )
    if (value == null) {
      logger(`Unable to set ${key} to ${value} as its null`);
      return;
    }

    // Example key = "j0.limitAdj"
    if (key.includes('.')) {
      const [joint, param] = key.split('.');

      // Update the config
      this.config[joint][param] = value;

      // Update the motor
      this.motors[joint][param] = value;
    } else {
      this.config[key] = value;
    }

    // Now write the config out
    if (save) this.writeConfig();

    logger(`updated config`, this.config);

    this.emit('meta');
  }

  /** ------------------------------
   * get state
   */
  get state() {
    // Build motors state object
    const motors = {};
    this.motors.forEach((motor) => {
      motors[motor.id] = motor.state;
    });

    // return state
    return {
      id: this.id,
      motors,
    };
  }

  /** ------------------------------
   * get meta
   */
  get meta() {
    // Build motors meta object
    const motors = {};
    this.motors.forEach((motor) => {
      motors[motor.id] = { id: motor.id };
    });

    // return meta
    return {
      stopped: this.stopped,
      ready: this.ready,
      home: this.home,
      homing: this.homing,
      moving: this.moving,
      config: this.config,
      motors,
      errors: this.errors,
    };
  }

  /** ------------------------------
   * validate
   *
   * validates to make sure action is ok based on parameters passed
   *
   * Example validate({ enabled: true, cleared: true, message: 'attempting to move '})
   * will prevent the action if the robot is not enabled and cleared of any errors
   */
  validate({ enabled, cleared, log }) {
    // If action requires robot to be enabled and we are not then error out
    if (enabled && this.stopped) {
      const message = `Please enable before ${log}`;
      logger(message);
      this.errors.push({ type: 'warning', message });
      this.emit('meta');
      return false;
    }
    // If action requires robot to have zero errors and we have errors then error out
    if (cleared && this.errors.length) {
      const message = `Please clear errors before ${log}`;
      logger(message);
      this.errors.push({ type: 'warning', message });
      this.emit('meta');
      return false;
    }
    return true;
  }

  /* -------------------- Motor Events -------------------- */

  motorHomed(id) {
    logger(`motor ${id} is homed`);

    // If we are homing robot check to see if we are all done homing
    if (this.homing) {
      if (this.motors.every((motor) => !motor.homing)) {
        logger(`all motors are home!`);
        this.homing = false;
      }
    }
    this.emit('meta');
    this.emit('state');
  }

  robotState() {
    this.emit('state');
  }

  robotEncoder() {
    this.emit('encoder');
  }

  motorMoved(id) {
    logger(`motor ${id} moved`);

    // If we are moving robot to a position check if its done
    if (this.moving) {
      if (this.motors.every((motor) => !motor.moving)) {
        logger(`all motors have moved!`);
        this.moving = false;
        this.emit('moved');
      }
    }

    // Let others know
    this.emit('meta');
    this.emit('state');
  }

  /* -------------------- Robot Actions -------------------- */

  robotHome() {
    logger(`home robot`);

    // Validate action
    if (!this.validate({ enabled: true, cleared: true, log: 'attempting to home' })) return;

    // Update our state
    this.homing = true;
    this.moving = true;

    // Because a motor might complete faster than all the others start we need to make sure we initialize all to homing + moving ;)
    this.motors.forEach((motor, i) => {
      motor.homing = true;
      motor.moving = true;
    });

    // Home all motors
    this.motors.forEach((motor, i) => {
      motor.goHome();
    });

    this.emit('meta');
  }

  robotCalibrate() {
    logger(`calibrate robot`);
  }

  robotSplitHome() {
    logger(`split home robot`);
  }

  robotStop() {
    logger(`stop robot`);

    this.stopped = true;

    // Disable all motors
    this.motors.forEach((motor) => {
      motor.disable();
    });

    this.emit('meta');
  }

  robotFreeze() {
    logger(`freeze robot`);

    // Freeze all motors ( stops but does not disable )
    this.motors.forEach((motor) => {
      motor.freeze();
    });

    this.emit('meta');
  }

  robotCenter() {
    logger(`center robot`);

    // We are moving whole robot
    this.moving = true;

    // Centers all motors
    this.motors.forEach((motor) => {
      motor.center();
    });

    this.emit('meta');
  }

  robotEnable() {
    logger(`enable robot`);

    this.stopped = false;

    // Enable all motors
    this.motors.forEach((motor) => {
      motor.enable();
    });

    this.emit('meta');
  }

  robotReset() {
    logger(`resetting robot errors`);

    // Reset all motors errors
    this.motors.forEach((motor) => {
      motor.resetErrors();
    });

    // Reset all robot errors
    this.errors = [];

    this.emit('meta');
  }

  robotSetAngles(angles, speed) {
    logger(`robotSetAngles at speed ${speed} angles:`, angles);

    // Validate action
    if (!this.validate({ enabled: true, cleared: true, log: 'attempting to move' })) return;

    // We are moving to a new location
    this.moving = true;

    // Because a motor might complete faster than all the others start we need to make sure we initialize all to moving ;)
    this.motors.forEach((motor, i) => {
      motor.moving = true;
    });

    // Step1: Determine travelSpeed and acceleration
    // Normally this involves complex math to ensure all motors start / stop at same time
    const travelSpeed = speed;
    const acceleration = speed;

    // Step2: Move via speed for each based on time
    this.motors.forEach((motor, i) => {
      logger(`setting angle for motor ${motor.id} at speed ${speed} and angle:`, angles[i]);
      motor.setPosition(angles[i], travelSpeed, acceleration);
    });

    this.emit('meta');
  }

  /* -------------------- Motor Actions -------------------- */

  motorSetPosition(id, pos, speed) {
    logger(`set position for motor ${id}`);

    // Skip if the motor is disabled
    if (!this.motorMap[id].enabled) {
      logger(`Not moving motor ${id}, please enable before attempting to move`);
      return;
    }

    this.motorMap[id].setPosition(pos, speed);
  }

  motorHome(id) {
    logger(`home motor ${id}`);
    this.motorMap[id].goHome();
  }

  motorResetErrors(id) {
    logger(`reset motor errors for motor ${id}`);
    this.motorMap[id].resetErrors();
  }

  motorEnable(id) {
    logger(`enable motor ${id}`);
    this.motorMap[id].enable();
  }

  motorDisable(id) {
    logger(`enable motor ${id}`);
    this.motorMap[id].disable();
  }

  motorZero(id) {
    logger(`zero motor ${id}`);
    this.motorMap[id].zero();
  }

  /* -------------------- Gripper Actions -------------------- */

  gripperSetPosition(pos, speed = 500) {
    logger(`set position for gripper to ${pos}, at speed ${speed}`);
  }
}

File: example_py_7/robot.py
import json
import threading
from pathlib import Path
from pyee import EventEmitter
from motor import Motor
from debug import Debug

logger = Debug('mock:robot\t')


class Robot(EventEmitter):
    def __init__(self, config):
        logger(f'Creating robot with id {config["id"]}')

        # Because we are event emitter
        super().__init__()

        self.id = config['id']  # id of the robot
        self.stopped = True  # if the robot is currently stopped
        self.ready = False  # if the robot is ready
        self.homing = False  # if the robot is currently homing
        self.home = False  # if the robot is home
        self.moving = False  # if the robot is moving
        self.freedrive = False  # if the robot is currently in freedrive
        self.motor_map = {}  # tracks all motors by joint id
        self.motors = []  # array of all motors (used for quick iteration)
        self.errors = []  # array of any errors that got triggered

        # Start up the robot when we are ready (normally you would be connecting to some control interface and waiting here)
        threading.Timer(2.0, self.setup).start()

    def setup(self):
        # First read in the config
        self.read_config()

        logger(f'Starting robot with id {self.id}')

        # Create the motors
        self.motor_map['j0'] = Motor(id='j0', **self.config['j0'])
        self.motor_map['j1'] = Motor(id='j1', **self.config['j1'])
        self.motor_map['j2'] = Motor(id='j2', **self.config['j2'])
        self.motor_map['j3'] = Motor(id='j3', **self.config['j3'])
        self.motor_map['j4'] = Motor(id='j4', **self.config['j4'])
        self.motor_map['j5'] = Motor(id='j5', **self.config['j5'])
        self.motor_map['j6'] = Motor(id='j6', **self.config['j6'])

        # Set motors into array
        self.motors = list(self.motor_map.values())

        # Subscribe to events for all motors
        for motor in self.motors:
            motor.on('homing', self.robot_state)
            motor.on('home', self.motor_homed)
            motor.on('disabled', self.robot_state)
            motor.on('freeze', self.robot_state)
            motor.on('enabled', self.robot_state)
            motor.on('moved', self.motor_moved)
            motor.on('reset', self.robot_state)

        # Create Gripper if you have a gripper

        # Report all encoder updates at 100ms interval
        self.encoder_timer = threading.Timer(0.1, self.robot_encoder)
        self.encoder_timer.start()

        self.ready = True
        self.emit('ready')

    def read_config(self):
        # Read in config file (create if it does not exist yet)
        try:
            # Get filename
            filename = Path('config.json')

            # Check if it exists and create if it does not
            if not filename.exists():
                logger('Config file does not exist, creating')
                with open(filename, 'w') as f:
                    json.dump({
                        'j0': {'limNeg': -180, 'limPos': 180, 'homePos': 0},
                        'j1': {'limNeg': -140, 'limPos': 80, 'homePos': 0},
                        'j2': {'limNeg': -140, 'limPos': 80, 'homePos': 0},
                        'j3': {'limNeg': -180, 'limPos': 180, 'homePos': 0},
                        'j4': {'limNeg': -95, 'limPos': 95, 'homePos': 0},
                        'j5': {'limNeg': -180, 'limPos': 180, 'homePos': 0}
                    }, f)

            # Read in config file
            with open(filename, 'r') as f:
                self.config = json.load(f)

            logger('Successfully read in config', self.config)
        except Exception as e:
            logger(f'Error reading config: {e}')

    def write_config(self):
        logger('Writing config to file', self.config)
        try:
            # Get filename
            filename = Path('config.json')
            # Write config
            with open(filename, 'w') as f:
                json.dump(self.config, f)
        except Exception as e:
            logger(f'Error writing config: {e}')

    def update_config(self, key, value, save=False):
        logger(f'Updating config {key} to {value}')

        # Special check (don't let user set a config param to null!!)
        if value is None:
            logger(f'Unable to set {key} to {value} as it is null')
            return

        # Example key = "j0.limitAdj"
        if '.' in key:
            joint, param = key.split('.')
            # Update the config
            self.config[joint][param] = value
            # Update the motor
            self.motor_map[joint][param] = value
        else:
            self.config[key] = value

        # Now write the config out
        if save:
            self.write_config()

        logger('Updated config', self.config)
        self.emit('meta')

    @property
    def state(self):
        # Build motors state object
        motors = {motor.id: motor.state for motor in self.motors}
        return {
            'id': self.id,
            'motors': motors,
            'tcpPose': [0, 0, 0, 0, 0, 0]
        }

    @property
    def busy(self):
        # Return true if we are busy doing anything
        return self.homing or self.moving

    @property
    def meta(self):
        # Build motors meta object
        motors = {motor.id: {'id': motor.id} for motor in self.motors}
        return {
            'stopped': self.stopped,
            'ready': self.ready,
            'home': self.home,
            'homing': self.homing,
            'moving': self.moving,
            'config': self.config,
            'motors': motors,
            'errors': self.errors,
            'freedrive': self.freedrive,
            'busy': self.busy,
        }

    def validate(self, enabled=False, cleared=False, log=''):
        # If action requires robot to be enabled and we are not then error out
        if enabled and self.stopped:
            message = f'Enable before {log}'
            logger(message)
            self.errors.append({'type': 'warning', 'message': message})
            self.emit('meta')
            return False
        # If action requires robot to have zero errors and we have errors then error out
        if cleared and self.errors:
            message = f'Clear errors before {log}'
            logger(message)
            self.errors.append({'type': 'warning', 'message': message})
            self.emit('meta')
            return False
        return True

    def motor_homed(self, id):
        logger(f'Motor {id} is homed')
        # If we are homing robot check to see if we are all done homing
        if self.homing:
            if all(not motor.homing for motor in self.motors):
                logger('All motors are home!')
                self.homing = False
        self.emit('meta')
        self.emit('state')

    def robot_state(self):
        self.emit('state')

    def robot_encoder(self):
        self.emit('encoder')
        self.encoder_timer = threading.Timer(0.1, self.robot_encoder)
        self.encoder_timer.start()

    def motor_moved(self, id):
        logger(f'Motor {id} moved')
        # If we are moving robot to a position check if it's done
        if self.moving:
            if all(not motor.moving for motor in self.motors):
                logger('All motors have moved!')
                self.moving = False
                self.emit('moved')
        # Let others know
        self.emit('meta')
        self.emit('state')

    def robot_home(self):
        logger('Home robot')
        # Validate action
        if not self.validate(enabled=True, cleared=True, log='attempting to home'):
            return
        # Update our state
        self.homing = True
        self.moving = True
        # Because a motor might complete faster than all the others start we need to make sure we initialize all to homing + moving ;)
        for motor in self.motors:
            if motor.enabled:
                motor.homing = True
                motor.moving = True
        # Home all motors
        for motor in self.motors:
            motor.go_home()
        self.emit('meta')

    def robot_zero(self):
        logger('Zero robot')
        # Zero out all motors
        for motor in self.motors:
            motor.zero()
        self.emit('meta')

    def robot_zero_ft(self):
        logger('Zero Force Tourque robot')
        self.emit('meta')

    def robot_calibrate(self):
        logger('Calibrate robot')

    def robot_split_home(self):
        logger('Split home robot')

    def robot_stop(self):
        logger('Stop robot')
        self.stopped = True
        # Disable all motors
        for motor in self.motors:
            motor.disable()
        self.emit('meta')

    def robot_freeze(self):
        logger('Freeze robot')
        # Freeze all motors (stops but does not disable)
        for motor in self.motors:
            motor.freeze()
        self.emit('meta')

    def robot_center(self):
        logger('Center robot')
        # We are moving the whole robot
        self.moving = True
        # Centers all motors
        for motor in self.motors:
            motor.center()
        self.emit('meta')

    def robot_enable(self):
        logger('Enable robot')
        self.stopped = False
        # Enable all motors
        for motor in self.motors:
            motor.enable()
        self.emit('meta')

    def robot_reset(self):
        logger('Resetting robot errors')
        # Reset all motors errors
        for motor in self.motors:
            motor.reset_errors()
        # Reset all robot errors
        self.errors = []
        self.emit('meta')

    def robot_set_angles(self, angles, speed):
        logger(f'robotSetAngles at speed {speed} angles: {angles}')
        # Validate action
        if not self.validate(enabled=True, cleared=True, log='attempting to move'):
            return
        # We are moving to a new location
        self.moving = True
        # Because a motor might complete faster than all the others start we need to make sure we initialize all to moving ;)
        for motor in self.motors:
            if motor.enabled:
                motor.moving = True

        # Step1: Determine travelSpeed and acceleration
        travel_speed = speed
        acceleration = speed
        # Step2: Move via speed for each based on time
        for motor, angle in zip(self.motors, angles):
            logger(
                f'Setting angle for motor {motor.id} at speed {speed} and angle: {angle}')
            motor.set_position(angle, travel_speed, acceleration)
        self.emit('meta')

    def robot_freedrive_enable(self, frame, cartFloatingAxis, nullspace=False):
        logger('Enabling freedrive')
        self.freedrive = True
        self.emit('meta')

    def robot_freedrive_disable(self):
        logger('Disabling freedrive')
        self.freedrive = False
        self.emit('meta')

    def motor_set_position(self, id, pos, speed):
        logger(f'Set position for motor {id}')
        self.motor_map[id].set_position(pos, speed)

    def motor_home(self, id):
        logger(f'Home motor {id}')
        self.motor_map[id].go_home()

    def motor_reset_errors(self, id):
        logger(f'Reset motor errors for motor {id}')
        self.motor_map[id].reset_errors()

    def motor_enable(self, id):
        logger(f'Enable motor {id}')
        self.motor_map[id].enable()

    def motor_disable(self, id):
        logger(f'Disable motor {id}')
        self.motor_map[id].disable()

    def motor_zero(self, id):
        logger(f'Zero motor {id}')
        self.motor_map[id].zero()

    def gripper_set_position(self, pos, speed=500, force=0):
        logger(
            f'Set position for gripper to {pos}, at speed {speed} with force {force}')
